src/client/Microsoft.Identity.Client/Instance/GenericAuthority.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using Microsoft.IdentityModel.Protocols;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+
+namespace Microsoft.Identity.Client.Instance
+{
+    internal class GenericAuthority : Authority
+    {
+        internal const string OpenIdConfigurationEndpointSuffix = ".well-known/openid-configuration";
+        private static readonly ConcurrentDictionary<string, OpenIdConnectConfiguration> s_openIdConnectConfigurations = new();
+           
+        internal GenericAuthority(AuthorityInfo authorityInfo)
+            : base(authorityInfo)
+        {
+
+        }
+
+        internal override string TenantId => null;
+    
+        internal IDocumentRetriever DocumentRetriever { get; set; } = new HttpDocumentRetriever();
+
+        internal override string GetTenantedAuthority(string tenantId, bool forceTenantless = false)
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override string GetTokenEndpoint()
+        {
+            string cacheKey = AuthorityInfo.CanonicalAuthority.ToString();
+            var configuration = s_openIdConnectConfigurations.GetOrAdd(cacheKey, RetrieveOpenIdConnectConfiguration);
+            return configuration.TokenEndpoint;
+        }
+
+        private OpenIdConnectConfiguration RetrieveOpenIdConnectConfiguration(string canonicalizedAuthority)
+        {
+            var configuration = OpenIdConnectConfigurationRetriever.GetAsync(AuthorityInfo.DiscoveryDocumentAddress, DocumentRetriever, CancellationToken.None)
```

Comment:
Thoughts about having this operation done in a critical section? To prevent a bunch of requests that are fired at the same time to make many discovery requests? 

We do this elsewhere like:
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/a07da18fd6a8ff1340fe9c7ca833c589ad796c3e/src/client/Microsoft.Identity.Client/Instance/Region/RegionManager.cs#L173

Reply:
implemented in https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3948/commits/a6454b41b05f539ea89bb5e5bdf12f54eb4651dc